### ISSUE TITLE
Getting cached thumbnail from kvstore is really slow

### ISSUE THREAD

**Issue by Gwildor** [jazzband/sorl-thumbnail#232]
Hello there!

We use this package for a project which has a lot of pages with a lot of small thumbnails on each of them. Because we were having performance issues after deploying to a test server, I started on analysing one of our heaviest page with the Django debug toolbar and the Python line profiler.

This particular page also has a lot of thumbnails, depending on the length of the dynamic content (few many-to-many relations which are shown as lists, each with a thumbnail, etc.). Currently I'm testing with data which causes the page to request 29 thumbnails from Sorl.

All calls are going through a custom template tag, which does some fancy media stuff and `img` tag rendering (actually through Python string formatting, because Django's templating language proofed to be too slow for this task), and uses the low-level `get_thumbnail` function (`from sorl.thumbnail import get_thumbnail`). Because I noticed that our custom template tag took about 150-200ms per call, I started my search there, and after some optimization on our end (such as removing the template rendering), the biggest bottleneck now is the `get_thumbnail` function (currently the template tag is down to 70-90ms per call).

I added line profiling on the default `ThumbnailBackend` and its `get_thumbnail` method, and I noticed that the line `cached = default.kvstore.get(thumbnail)` ( [link to code](https://github.com/mariocesar/sorl-thumbnail/blob/0d53f8a2/sorl/thumbnail/base.py#L92) ) was taking about 45ms per call when running version 11.12 of this package. I hoped this was purely because we did not implement a caching backend yet, so I added memcached, and tried again. Unfortunately, the call to that particularly line was still around 45ms per call (actually with surprisingly little divergence between the calls), but perhaps this was because the calls were now going through my Vagrant instance (I was running tests without Vagrant before installing memcached). Please note I was running memcached in verbose mode (`memcached -vv`) to make sure Sorl was using memcached correctly.

So then I decided to install the latest version from this repository ( 0d53f8a2 ) and try again. I immediately noticed a change, but unfortunately not in a good way. This time around, each call took over twice as long as it did before. Instead of a steady 45ms per call to retrieve the cached thumbnail, it now took at least 100ms but averaging around 130ms, and with peaks up to 185ms, over 4(!) times as long as it took before. However, this, again, was without using memcached. So then I started memcached again, and the results were the same. But then I ran it in verbose mode again, to double check, and suddenly the times started dropping again, all the way down to 39ms, but still with peaks up to 83ms). I have almost no experience with memcached before today, so I'm probably doing something wrong with starting it, because when running it as a service it also still seems to do SQL queries, which it isn't doing when running it verbose.

So, all in all, my findings of this line in the code:

| Version | default DB cache | `service memcached start` | `memcached -vv` |
| --- | --- | --- | --- |
| 11.12 | ±45ms | ±45ms | ±45ms |
| 0d53f8a2 | 100-180ms | 100-180ms | 39-83ms |

So, in conclusion:
- Getting the cached thumbnail is way slower now than it used to be in 11.12.
- It was already way too slow (in my humble opinion) in 11.12 to begin with. When you have 20 thumbnails on the page, which I agree is high but not that extreme, you already have +1s response time.
- I guess I'm starting memcached wrong, so I think it's safe to ignore the penultimate column, but I'm just including it for completeness.
